modules/isd/src/ResidueProteinProximityRestraint.cpp

```diff
@@ -0,0 +1,159 @@
+/**
+ *  \file isd/ResidueProteinProximityRestraint.h
+ *  \brief Restraint a selection of particles (eg. a residue or 
+ *  segment) to be within  a certain distance of a second 
+ *  selection of particles (eg. a protein). 
+ *  Use to model data from mutagenesis experiments that disrupt
+ *  protein-protein interactions.
+ *
+ *  Copyright 2007-2022 IMP Inventors. All rights reserved.
+ *
+ */
+
+#include <IMP/isd/ResidueProteinProximityRestraint.h>
+#include <IMP/algebra/VectorD.h>
+#include <IMP/core/XYZ.h>
+#include <IMP/isd/Scale.h>
+#include <IMP/container/ListPairContainer.h>
+#include <boost/math/special_functions/erf.hpp>
+#include <boost/unordered_map.hpp>
+#include <boost/tuple/tuple.hpp>
+#include <math.h>
+
+IMPISD_BEGIN_NAMESPACE
+
+ResidueProteinProximityRestraint::ResidueProteinProximityRestraint(Model* m,
+						 Float cutoff,
+						 Float sigma,
+                                                 Float xi,
+						 bool part_of_log_score,
+						 std::string name):
+  Restraint(m,name),
+  cutoff_(cutoff),
+  sigma_(sigma),
+  xi_(xi),
+  part_of_log_score_(part_of_log_score)
+  {
+}
+
+// add a contribution: general case
+void ResidueProteinProximityRestraint::add_pairs_container(PairContainer *pc) {
+  contribs_.push_back(pc);
+  default_range_.push_back((int)default_range_.size());
+}
+
+void ResidueProteinProximityRestraint::add_contribution_particles(const ParticleIndexes ppis1,
+                                                         const ParticleIndexes ppis2){
+
+  ParticlesTemp pps1;
+  ParticlesTemp pps2;
+  for (unsigned int i = 0; i < ppis1.size(); i++){
+    pps1.push_back(get_model()->get_particle(ppis1[i]));
+    ppis_.push_back(ppis1[i]);
+  }
+  for (unsigned int i = 0; i < ppis2.size(); i++){
+    pps2.push_back(get_model()->get_particle(ppis2[i]));
+    ppis_.push_back(ppis2[i]);
+  }
+  
+  IMP::core::XYZRs xx1 = IMP::core::XYZRs(pps1);
+  IMP::core::XYZRs xx2 = IMP::core::XYZRs(pps2);
+
+  algebra::Vector3D centroid1 = core::get_centroid(xx1);
+  algebra::Vector3D centroid2 = core::get_centroid(xx2);
+  
+  coms1_.push_back(xx1);
+  coms2_.push_back(xx2);
+  //sigmass_.push_back(sigmas);
+}
+
+double ResidueProteinProximityRestraint::evaluate_for_contributions(Ints c)
+  const{
+
+  double score_tot = 0.0;
+  double score = 0.0;
+  //! Loop over the contributions and score things
+  for (Ints::const_iterator nit=c.begin();nit!=c.end();++nit){
+    int n = *nit;
```

Comment:
With C++11 this can be written more concisely as `for (int n : c) {`

Reply:
Done